### PR TITLE
[lang] Add spmv and sparse linear solver using ndarray as vector

### PR DESCRIPTION
**Pull request by FantasyVR**
Issue: #2906 

Previously, ndarray cann't be used as a vector on CPU backend in spmv and sparse linear solve operations. In this PR, the ndarray is [mapped](https://eigen.tuxfamily.org/dox/classEigen_1_1Map.html) as Eigen's vector, and then it's used to do math computation. 

In addition, the implicit mass spring example is modified to use ndarrays. It enables us to run this example on both the CPU and GPU backend.

### CODE REVIEW

File: taichi/program/sparse_matrix.cpp [hunk 1]
```diff
@@ -24,6 +24,12 @@
         }                                                                      \
   }
 
+#define INSTANTIATE_SPMV(type, storage)                               \
+  template void                                                       \
+  EigenSparseMatrix<Eigen::SparseMatrix<type, Eigen::storage>>::spmv( \
+      taichi::lang::Program *prog, const taichi::lang::Ndarray &x,    \
```
Comment (strongoier) [inline taichi/program/sparse_matrix.cpp:30]: Why is `x` const but not `y`?

Comment (FantasyVR) [reply taichi/program/sparse_matrix.cpp:30]: `y` is used to store the computed results of `A @ x = y`. `x` is the input of sparse matrix-vector multiplication.

Comment (strongoier) [reply taichi/program/sparse_matrix.cpp:30]: Hmm but you are not modifying the `Ndarray` object. You are just writing to the memory address `y` wraps.

File: python/taichi/examples/simulation/implicit_mass_spring.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+# pylint: disable=E1101
```
[the remark this answers is not in the document]
Comment (FantasyVR) [reply python/taichi/examples/simulation/implicit_mass_spring.py:1]: If `E1101` is not disabled, this example will fail to pass `pylint`.

```
pylint...................................................................Failed
- hook id: pylint
- exit code: 2

************* Module implicit_mass_spring
python/taichi/examples/simulation/implicit_mass_spring.py:9:1: E1101: Module 'taichi' has no 'data_oriented' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:16:19: E1101: Module 'taichi' has no 'Vector' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:16:38: E1101: Module 'taichi' has no 'f32' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:17:23: E1101: Module 'taichi' has no 'Vector' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:17:42: E1101: Module 'taichi' has no 'f32' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:18:19: E1101: Module 'taichi' has no 'Vector' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:18:38: E1101: Module 'taichi' has no 'f32' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:19:21: E1101: Module 'taichi' has no 'Vector' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:19:40: E1101: Module 'taichi' has no 'f32' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:20:20: E1101: Module 'taichi' has no 'field' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:20:29: E1101: Module 'taichi' has no 'f32' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:22:22: E1101: Module 'taichi' has no 'Vector' member (no-member)
python/taichi/examples/simulation/implicit_mass_spring.py:22:41: E1101: Module 'taichi' has no 'i32' member (no-member)
...
```

**Review by Hanke98 — APPROVED** [review]
LGTM! An off-topic concern is that we are using too many macros especially in [pr#6657](https://github.com/taichi-dev/taichi/pull/6657). Although the macros improve the code reuse,   they might stop potential users to go deeper in our implementation by reading and understanding the code. So my suggestion is adding some comments on those macros for now and try to refactor them into inline functions later. :-)

File: python/taichi/examples/simulation/implicit_mass_spring.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (strongoier) [reply python/taichi/examples/simulation/implicit_mass_spring.py:1]: Is this the CI result or your local result? There seems to be some configuration problem.

[the remark this answers is not in the document]
Comment (FantasyVR) [reply python/taichi/examples/simulation/implicit_mass_spring.py:1]: It's the local results. With `pre-commit` enabled, I couldn't commit `implicit_mass_spring` changes. I can try to force pushing and see if CI past.

**Review by strongoier — APPROVED** [review]

**Pull request merged**